File: backend/app/api/diagnosis.py
"""
Diagnosis API - Image upload and acne severity classification
"""

from fastapi import APIRouter, UploadFile, File, Form, HTTPException, Depends
from sqlalchemy.orm import Session
from sqlalchemy import select
from pydantic import BaseModel
from typing import Optional, Dict, List
import os
import uuid
import json
from datetime import datetime

from app.ml.acne_classifier import AcneClassifier
from app.ml.multimodal_fusion import MultimodalFusion
from app.core.dependencies import get_current_user
from app.core.database import get_db
from app.core.config import settings
from app.models.db_models import User, Diagnosis as DiagnosisModel

router = APIRouter()

# Initialize ML models (lazy loading)
_classifier = None
_fusion = None


def get_classifier():
    global _classifier
    if _classifier is None:
        _classifier = AcneClassifier()
    return _classifier


def get_fusion():
    global _fusion
    if _fusion is None:
        _fusion = MultimodalFusion()
    return _fusion


class ClinicalMetadata(BaseModel):
    age: int
    skin_type: str = "normal"
    acne_duration_months: int = 6
    previous_treatments: Optional[List[str]] = None
    allergies: Optional[List[str]] = None


class DiagnosisResponse(BaseModel):
    id: str
    severity: str
    confidence: float
    severity_scores: Dict[str, float]
    lesion_counts: Dict[str, int]
    affected_areas: List[str]
    clinical_notes: str
    recommended_urgency: str
    image_url: str
    created_at: str


@router.post("/analyze", response_model=DiagnosisResponse)
async def analyze_skin_image(
    image: UploadFile = File(..., description="Skin image for analysis"),
    clinical_metadata: str = Form(..., description="JSON string of clinical metadata"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Analyze skin image for acne severity classification.
    
    Upload a skin image with clinical metadata to get AI-powered diagnosis.
    """
    # Parse metadata
    try:
        metadata = json.loads(clinical_metadata)
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail="Invalid clinical metadata JSON")
    
    # Validate image
    if not image.content_type or not image.content_type.startswith("image/"):
        raise HTTPException(status_code=400, detail="File must be an image")
    
    # Read image contents
    contents = await image.read()
    
    # Validate file size
    if len(contents) > settings.MAX_UPLOAD_SIZE:
        max_size_mb = settings.MAX_UPLOAD_SIZE / (1024 * 1024)
        raise HTTPException(
            status_code=413,
            detail=f"File size exceeds maximum allowed size of {max_size_mb}MB"
        )
    
    # Save image
    ext = image.filename.split(".")[-1] if image.filename else "jpg"
    filename = f"{uuid.uuid4()}.{ext}"
    filepath = os.path.join(settings.UPLOAD_DIR, filename)
    
    # Ensure upload directory exists
    os.makedirs(settings.UPLOAD_DIR, exist_ok=True)
    
    with open(filepath, "wb") as f:
        f.write(contents)
    
    # Get ML models
    classifier = get_classifier()
    
    # Check if using Keras model (predictions) or PyTorch (features)
    if classifier.use_keras:
        # For Keras: Get predictions directly from model (no fusion needed for pure model output)
        print("🔍 Using Keras model - getting direct predictions")
        model_predictions = classifier.extract_features(filepath)  # This returns predictions, not features
        classification = classifier.classify(model_predictions, image_path=filepath)
    else:
        # For PyTorch: Use multimodal fusion
        print("🔍 Using PyTorch model - using multimodal fusion")
        fusion = get_fusion()
        image_features = classifier.extract_features(filepath)
        fused_features = fusion.fuse(image_features, metadata)
        classification = classifier.classify(fused_features, image_path=filepath)
    
    # Detect lesions (pass severity to ensure consistency)
    lesion_counts = classifier.detect_lesions(filepath, classification["severity"])
    
    # Generate clinical notes
    clinical_notes = generate_clinical_notes(
        classification["severity"],
        lesion_counts,
        metadata
    )
    
    # Create diagnosis record
    diagnosis_id = str(uuid.uuid4())[:8]
    diagnosis = DiagnosisModel(
        id=diagnosis_id,
        user_id=current_user.id,
        severity=classification["severity"],
        confidence=int(classification["confidence"] * 100),
        severity_scores=classification["all_scores"],
        lesion_counts=lesion_counts,
        affected_areas=classification.get("affected_areas", ["face"]),
        clinical_notes=clinical_notes,
        recommended_urgency=get_urgency(classification["severity"]),
        image_url=f"/uploads/{filename}",
        clinical_metadata=metadata,
    )
    
    db.add(diagnosis)
    db.commit()
    db.refresh(diagnosis)
    
    return DiagnosisResponse(
        id=diagnosis.id,
        severity=diagnosis.severity,
        confidence=diagnosis.confidence / 100.0,
        severity_scores=diagnosis.severity_scores,
        lesion_counts=diagnosis.lesion_counts,
        affected_areas=diagnosis.affected_areas,
        clinical_notes=diagnosis.clinical_notes,
        recommended_urgency=diagnosis.recommended_urgency,
        image_url=diagnosis.image_url,
        created_at=diagnosis.created_at.isoformat()
    )


@router.get("/{diagnosis_id}")
async def get_diagnosis(
    diagnosis_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get diagnosis by ID (user-specific)."""
    diagnosis = db.execute(
        select(DiagnosisModel).where(
            DiagnosisModel.id == diagnosis_id,
            DiagnosisModel.user_id == current_user.id
        )
    ).scalar_one_or_none()
    
    if not diagnosis:
        raise HTTPException(status_code=404, detail="Diagnosis not found")
    
    return {
        "id": diagnosis.id,
        "severity": diagnosis.severity,
        "confidence": diagnosis.confidence / 100.0,
        "severity_scores": diagnosis.severity_scores,
        "lesion_counts": diagnosis.lesion_counts,
        "affected_areas": diagnosis.affected_areas,
        "clinical_notes": diagnosis.clinical_notes,
        "recommended_urgency": diagnosis.recommended_urgency,
        "image_url": diagnosis.image_url,
        "created_at": diagnosis.created_at.isoformat(),
        "metadata": diagnosis.clinical_metadata
    }


@router.get("/")
async def list_diagnoses(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """List all diagnoses for current user."""
    diagnoses = db.execute(
        select(DiagnosisModel).where(DiagnosisModel.user_id == current_user.id)
        .order_by(DiagnosisModel.created_at.desc())
    ).scalars().all()
    
    return [
        {
            "id": d.id,
            "severity": d.severity,
            "confidence": d.confidence / 100.0,
            "severity_scores": d.severity_scores,
            "lesion_counts": d.lesion_counts,
            "affected_areas": d.affected_areas,
            "clinical_notes": d.clinical_notes,
            "recommended_urgency": d.recommended_urgency,
            "image_url": d.image_url,
            "created_at": d.created_at.isoformat(),
            "metadata": d.clinical_metadata
        }
        for d in diagnoses
    ]


def generate_clinical_notes(severity: str, lesion_counts: dict, metadata: dict) -> str:
    """Generate clinical notes based on analysis."""
    notes = []
    
    total = sum(lesion_counts.values())
    
    # Generate severity-specific notes that are consistent with lesion counts
    if severity == "clear":
        if total == 0:
            notes.append("No significant acne lesions detected.")
        else:
            notes.append("Minimal acne lesions detected. Skin appears relatively clear.")
    elif severity == "mild":
        notes.append(f"Mild acne detected with {total} total lesions (primarily comedones and papules).")
    elif severity == "moderate":
        notes.append(f"Moderate acne with {total} total lesions including papules and pustules.")
    elif severity == "severe":
        notes.append(f"Severe acne with {total} total lesions including numerous pustules and nodules.")
    elif severity == "very_severe":
        notes.append(f"Very severe cystic acne with {total} total lesions including nodules and cysts. Requires aggressive treatment.")
    else:
        notes.append(f"Acne severity: {severity}. Total lesions detected: {total}.")
    
    # Add specific lesion breakdown if there are lesions
    if total > 0:
        lesion_details = []
        if lesion_counts.get("nodules", 0) > 0 or lesion_counts.get("cysts", 0) > 0:
            lesion_details.append("nodular/cystic lesions present")
        if lesion_counts.get("pustules", 0) > 5:
            lesion_details.append("multiple inflammatory pustules")
        if lesion_details:
            notes.append("Note: " + ", ".join(lesion_details) + ".")
    
    # Add duration note if chronic
    if metadata.get("acne_duration_months", 0) > 12:
        notes.append("Chronic acne (>12 months) - consider comprehensive treatment plan.")
    
    return " ".join(notes)


def get_urgency(severity: str) -> str:
    """Determine urgency based on severity."""
    urgency_map = {
        "clear": "routine",
        "mild": "routine", 
        "moderate": "soon",
        "severe": "soon",
        "very_severe": "urgent"
    }
    return urgency_map.get(severity, "routine")

File: backend-node/src/routes/diagnosis.ts
import { Router, Request, Response } from 'express';
import { body, validationResult } from 'express-validator';
import { authenticate } from '../middleware/auth';
import { upload, validateImageFile } from '../utils/fileUpload';
import { classify, predictType } from '../services/ml/inference';
import {
  createDiagnosis,
  getDiagnosisById,
  listDiagnoses,
  generateClinicalNotes,
  getUrgency,
  detectLesions,
} from '../services/diagnosisService';
import { join } from 'path';
import { config } from '../config';

const router = Router();

/**
 * POST /api/v1/diagnosis/analyze
 * Analyze skin image for acne severity and type classification
 */
router.post(
  '/analyze',
  authenticate,
  upload.single('image'),
  [
    body('clinical_metadata').isJSON().withMessage('clinical_metadata must be valid JSON'),
  ],
  async (req: Request, res: Response) => {
    try {
      // Validate file
      validateImageFile(req.file);

      // Parse metadata
      let metadata: Record<string, any>;
      try {
        metadata = JSON.parse(req.body.clinical_metadata);
      } catch (error) {
        return res.status(400).json({ detail: 'Invalid clinical metadata JSON' });
      }

      const user = req.user!;
      const file = req.file!;

      // Run ML inference
      const classification = await classify(file.path);
      if (!classification) {
        return res.status(500).json({ detail: 'ML inference failed' });
      }

      // Acne type classification (only if acne detected)
      let typeResult: { type: string; confidence: number; allScores: Record<string, number> } | null = null;
      if (classification.hasAcne) {
        typeResult = await predictType(file.path);
      }

      // Detect lesions
      const lesionCounts = detectLesions(classification.severity);

      // Generate clinical notes
      const clinicalNotes = generateClinicalNotes(
        classification.severity,
        lesionCounts,
        metadata
      );

      // Generate problem summary (simplified - can be enhanced later)
      const problemSummary = {
        severity: classification.severity,
        severity_index: classification.severityIndex,
        lesion_counts: lesionCounts,
        clinical_metadata: metadata,
        affected_areas: ['face'],
        confidence: classification.confidence,
      };

      // Create diagnosis record
      const diagnosis = await createDiagnosis(user.id, {
        severity: classification.severity,
        confidence: Math.round(classification.confidence * 100),
        severityScores: classification.allScores,
        lesionCounts,
        affectedAreas: ['face'],
        clinicalNotes,
        recommendedUrgency: getUrgency(classification.severity),
        imageUrl: `/uploads/${file.filename}`,
        clinicalMetadata: metadata,
        problemSummary,
      });

      // Return response
      res.json({
        id: diagnosis.id,
        has_acne: classification.hasAcne,
        binary_confidence: classification.binaryConfidence,
        severity: diagnosis.severity,
        confidence: diagnosis.confidence / 100.0,
        severity_scores: diagnosis.severityScores,
        acne_type: typeResult?.type || null,
        acne_type_confidence: typeResult?.confidence || null,
        acne_type_scores: typeResult?.allScores || null,
        lesion_counts: diagnosis.lesionCounts,
        affected_areas: diagnosis.affectedAreas,
        clinical_notes: diagnosis.clinicalNotes,
        recommended_urgency: diagnosis.recommendedUrgency,
        image_url: diagnosis.imageUrl,
        created_at: diagnosis.createdAt.toISOString(),
        problem_summary: diagnosis.problemSummary,
      });
    } catch (error: any) {
      console.error('Diagnosis error:', error);
      if (error.message === 'No file uploaded' || error.message.includes('File must be')) {
        return res.status(400).json({ detail: error.message });
      }
      if (error.message.includes('File size exceeds')) {
        return res.status(413).json({ detail: error.message });
      }
      res.status(500).json({ detail: 'Internal server error' });
    }
  }
);

/**
 * GET /api/v1/diagnosis/:id
 * Get diagnosis by ID
 */
router.get('/:id', authenticate, async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const user = req.user!;

    const diagnosis = await getDiagnosisById(id, user.id);
    if (!diagnosis) {
      return res.status(404).json({ detail: 'Diagnosis not found' });
    }

    res.json({
      id: diagnosis.id,
      severity: diagnosis.severity,
      confidence: diagnosis.confidence / 100.0,
      severity_scores: diagnosis.severityScores,
      lesion_counts: diagnosis.lesionCounts,
      affected_areas: diagnosis.affectedAreas,
      clinical_notes: diagnosis.clinicalNotes,
      recommended_urgency: diagnosis.recommendedUrgency,
      image_url: diagnosis.imageUrl,
      created_at: diagnosis.createdAt.toISOString(),
      metadata: diagnosis.clinicalMetadata,
    });
  } catch (error) {
    console.error('Get diagnosis error:', error);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

/**
 * GET /api/v1/diagnosis
 * List all diagnoses for current user
 */
router.get('/', authenticate, async (req: Request, res: Response) => {
  try {
    const user = req.user!;
    const diagnoses = await listDiagnoses(user.id);

    res.json(
      diagnoses.map((d) => ({
        id: d.id,
        severity: d.severity,
        confidence: d.confidence / 100.0,
        severity_scores: d.severityScores,
        lesion_counts: d.lesionCounts,
        affected_areas: d.affectedAreas,
        clinical_notes: d.clinicalNotes,
        recommended_urgency: d.recommendedUrgency,
        image_url: d.imageUrl,
        created_at: d.createdAt.toISOString(),
        metadata: d.clinicalMetadata,
      }))
    );
  } catch (error) {
    console.error('List diagnoses error:', error);
    res.status(500).json({ detail: 'Internal server error' });
  }
});

export default router;

